### PR TITLE
feat(paginate): Return exact types from paginate()

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

`paginate()` currently returns a naive `GetStaticPathsResult` which override the actual types it returns, this PR makes it so it actually returns what is being inputted, which because of the fact that it allows you to inject params and props, made this strangely hard. Maybe there's an easier way to do it, but I couldn't figure it out

This is similar to https://github.com/withastro/astro/pull/8048, but it doesn't require user to manually type `PaginateFunction`

This PR also fixes https://github.com/withastro/astro/issues/7159, mostly by accident

## Testing

Tested manually

## Docs

N/A. Docs document the return values of paginate(), so this really just makes it more accurate to what's currently described

### CODE REVIEW

File: packages/astro/client.d.ts [hunk 1]
```diff
@@ -57,7 +57,7 @@ declare module 'astro:assets' {
 	};
 
 	type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };
-	type Simplify<T> = { [KeyType in keyof T]: T[KeyType] };
+	type Simplify<T> = { [KeyType in keyof T]: T[KeyType] } & {};
```
Comment (Princesseuh) [inline packages/astro/client.d.ts:60]: Just a small fix, eventually when we refactor types, we should probably have a "internal-type-utils" file somewhere with this ~~and all the other nonsense I add~~

File: packages/astro/src/core/render/route-cache.ts [hunk 2]
```diff
@@ -50,7 +51,9 @@ export async function callGetStaticPaths({
 	// Calculate your static paths.
 	let staticPaths: GetStaticPathsResult = [];
 	staticPaths = await mod.getStaticPaths({
-		paginate: generatePaginateFunction(route),
+		// Q: Why the cast?
+		// A: So users downstream can have nicer typings, we have to make some sacrifice in our internal typings, which necessitate a cast here
+		paginate: generatePaginateFunction(route) as PaginateFunction,
```
Comment (Princesseuh) [inline packages/astro/src/core/render/route-cache.ts:56]: I think there is a way to make this work without the cast, but I couldn't for my life figure it out. I think the disconnect between the internal implementation and what the user actually uses makes it very hard to do in one type

Comment (natemoo-re) [reply packages/astro/src/core/render/route-cache.ts:56]: Seems like a reasonable tradeoff!

**Review by natemoo-re — APPROVED** [review]
Looks great, approved!

Would you mind closing out https://github.com/withastro/astro/pull/8048 and the related issue after merging this since they overlap?

**Review by matthewp — APPROVED** [review]

**Pull request merged**